test/spec/form.spec.js

```diff
@@ -1724,6 +1724,77 @@ describe('Form.prototype', () => {
     });
 });
 
+describe('Missing external instances', () => {
+    it('does not produce extraneous/unrelated errors', () => {
+        const options = { full: true };
+        const session = {
+            deviceid: 'a',
+            username: 'b',
+            email: 'c',
+            phonenumber: 'd',
+            simserial: 'e',
+            subscriberid: 'f',
+        };
+        const externalInstances = [undefined, null];
+        const form = loadForm(
+            'missing-external-instance.xml',
+            undefined,
+            options,
+            session,
+            externalInstances
+        );
+        const loadErrors = form.init();
+
+        loadErrors.forEach((error) => {
+            expect(error).not.to.include(
+                'FormLogicError: instance "__session" does not exist in model'
+            );
+            expect(error).not.to.include('Error trying to parse XML model');
+        });
+
+        expect(loadErrors).to.deep.equal([
+            "Can't find brands.csv.",
+            "Can't find products.csv.",
+        ]);
+    });
+
+    it('does not produce extraneous/unrelated errors', () => {
```

Comment:
I think this name is supposed to be different? Is it intended to show the empty attachment error? I’m surprised that the error is the same as above.

Reply:
Ahh, got it. So the first test has `undefined` and `null`. And this one has defined sources but no file. To me it would make more sense to either have all three in one form or three different tests. I think having all that `session` stuff also made me miss the difference between the two.

I think adding `when the value is missing` and `when there is a value but the data for the external instance is missing` would really help.

I don't know what a table-style test is. 😬

Reply:
> I don't know what a table-style test is. 😬

Ah sorry, I’m not even sure what the appropriate jargon for it is! Parameterized tests? I’m referring to looping over a set of fixtures to use in the test. Commonly used as `it.each` in other JS test frameworks, often as in a standard `[...].forEach`. I tend to avoid it unless it’s a sure thing that it won’t get branchy, because test helpers are hard to untangle once they exist.

I agree the changes you suggested make sense. I’ll get that pushed up tomorrow.

Reply:
Yes, yes, I'm with you all around!